worker/pubsub/remoteserver.go

```diff
@@ -0,0 +1,322 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/pubsub"
+	"github.com/juju/retry"
+	"github.com/juju/utils/clock"
+	"github.com/juju/utils/deque"
+	"launchpad.net/tomb"
+
+	"github.com/juju/juju/api"
+	pubsubapi "github.com/juju/juju/api/pubsub"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/pubsub/forwarder"
+	"github.com/juju/juju/worker"
+)
+
+// RemoteServer represents the public interace of the worker
+// responsible for forwarding messages to a single other API server.
+type RemoteServer interface {
+	worker.Worker
+	UpdateAddresses(addresses []string)
+	Publish(message *params.PubSubMessage)
+}
+
+// remoteServer is responsible for taking messages and sending them to the
+// pubsub endpoint on the remote server. If the connection is dropped, the
+// remote server will try to reconnect. Messages are not sent until the
+// connection either succeeds the first time, or fails to connect. Once there
+// is a failure, incoming messages are dropped until reconnection is complete,
```

Comment:
Why the decision to drop messages when there's no connection? I don't really understand the implications of this. Could there be a problem with silently dropping them?